File: 2108-StudentPortalService/student-portal-final/src/main/java/studentportal/model/Address.java
package studentportal.model;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;

import org.hibernate.annotations.NaturalId;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * Simple Address model that tries to mirror Google API and 
 *
 */
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "Addresses")
public final class Address {

	@Id
	@Column(name="address_id")
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private int addressId;
	
	@NaturalId
	@Column(name="place_id",unique = true, nullable = false)
	private String placeId;
	
	@Column(name="streetNumber")
	private String streetNumber;
	
	@Column(name="route")
	private String route;
	
	@Column(name="locality")
	private String locality;
	
	@Column(name="administrativeAreaLevel1")
	private String administrativeAreaLevel1;
	
	@Column(name="country")
	private String country;
	
	@Column(name="postalCode")
	private String postalCode;

	@Override
	public String toString() {
		return placeId+": " + streetNumber +" "+ route +" "+ locality +" "+ administrativeAreaLevel1 +" "+ country +" "+ postalCode;
	}

}

File: 2108-StudentPortalReact/student-portal-ts/src/util/address.ts
import { IAddress, IAddressComponents } from "../state-structures";

export function parseAddress(addressAPIArray: IAddressComponents[]): IAddress {
    let address: IAddress = {
        placeId: "",
        streetNumber: "",
        route: "",
        locality: "",
        administrativeAreaLevel1: "",
        country: "",
        postalCode: "",
    };

    addressAPIArray.forEach((addresscomponent) => {
        addresscomponent.types.forEach((type) => {
            switch (type) {
                case "street_number":
                    address.streetNumber = addresscomponent.short_name
                    break;
                case "postal_code":
                    address.postalCode = addresscomponent.short_name
                    break;
                case "route":
                    address.route = addresscomponent.short_name
                    break;
                case "locality":
                    address.locality = addresscomponent.short_name
                    break;
                case "administrative_area_level_1":
                    address.administrativeAreaLevel1 = addresscomponent.short_name
                    break;
                case "country":
                    address.country = addresscomponent.long_name
                    break;
                default:
                    break;
            }
        });
    });

    address.placeId = `${address.streetNumber} ${address.route}, ${address.locality}, ${address.administrativeAreaLevel1}, ${address.postalCode}, ${address.country}`
    address.placeId = address.placeId.trim();
    
    return address;
}
